File: 参考/comp9041-454/9041-labonserver/lab07/log_probability.pl
#!/usr/bin/perl -w
$word=$ARGV[0];
%cw=();
%ct=();
foreach $file (glob "lyrics/*.txt") {
  $len=0;
  $len2=0;
  open F,'<',"$file" or die;
  while ($line=<F>){
    @a=$line=~ /\b$word\b/ig;
    $len+=@a;
    @b=$line=~ /[a-zA-Z]+/g;
    $len2+=@b;
  }
  $file=~ s/.*\///;
  $file=~ s/.txt//;
  $file=~ s/[^a-z]/ /gi;
  $cw{$file}=$len;
  $ct{$file}=$len2;
}


for $i (sort keys %cw){
    printf "log((%d+1)/%6d) = %8.4f %s\n",$cw{$i},$ct{$i},log(($cw{$i}+1)/$ct{$i}),$i;
}

File: others/comp9041-454/9041-labonserver/lab07/log_probability.py
#!/usr/bin/python
import sys
import re
import glob
import math
word=sys.argv[1]
cw={}
ct={}
word_len=0
tot_len=0
file_list=glob.glob(r"lyrics/*.txt")
for filename in file_list:
    with open(filename) as file:
        data = file.read()
        list=re.findall(r'\b{}\b'.format(word),data,flags=re.IGNORECASE)
        word_len=len(list)
        list2=re.findall(r'[a-zA-Z]+',data)
        tot_len=len(list2)
        filename=re.sub(r'.*/','',filename)
        filename=re.sub(r'.txt','',filename)
        filename=re.sub(r'[^a-z]',' ',filename,flags=re.IGNORECASE)
        cw[filename]=word_len
        ct[filename]=tot_len

for i in sorted(cw.keys()):
    print "log((%d+1)/%6d) = %8.4f %s" % (cw[i],ct[i],math.log(float(cw[i]+1)/float(ct[i])),i)
